### ISSUE TITLE
Is it possible to collect producer batch request latency and error information?

### ISSUE THREAD

**Issue by beanliu** [twmb/franz-go#130]
The best I could get so far is `HookProduceBatchWritten` and `HookBrokerWrite`. But they are not perfect:

- `HookBrokerWrite`: impossible to distinguish if it's a produce request (`bytesWritten` is a strong signal but not always).
- `HookProduceBatchWritten`: lack of latency and error information.

Maybe I'm missing something in the docs. But it would be great to have request latency and error metrics for producer monitoring.

**Comment by twmb** [comment]
`HookBrokerE2E` can be used by processing the latency when `key == 0` (same thing for `HookBrokerWrite`, but E2E is a bit more robust in the latency). This would be the latency from writing the bytes to the wire through reading the response, but not processing it.

I could potentially add latency to `ProduceBatchMetrics` as well. This could be any form of latency: from writing through reading the bytes to _just before_ processing, or through processing (probably both options available). This would be an easier way to hook into latency, but would have a tiny bit of overhead from after the time the bytes are actually read to just before the data is processed.

So in summary:
- It's possible to get some latency today by hooking into the `key` field, as produce request is key 0. Using this would allow you to get latency today.
- This issue is a good feature request for adding more latency metrics to `ProduceBatchMetrics`.

I'm planning to spend some time this weekend on a recent batch of feature requests.

**Comment by beanliu** [comment]
💯  cool, thanks.

**Comment by twmb** [comment]
Looking at this further, I think it makes sense to keep in `BrokerE2E`, which is specific to the request lifecycle itself, vs `ProduceBatchMetrics`, which is specific to an individual batch within a produce request. Adding it to `ProduceBatchMetrics` may erroneously make it look like every batch in a request had a certain amount of latency, vs. the entire request itself and all batches within it having the one single latency.

Do you want per topic / per partition latency breakdown, or is per-request latency good? Per-record latency can currently be gleaned from the timestamp field within a record (set once buffered) and the time once the promise is called.

How would per-batch latency be used / beneficial in comparison to per-request and potentially per-record?

**Comment by beanliu** [comment]
Agree that adding latency info to `ProduceBatchMetrics` is confusing. And per-request latency is good enough in my case. However, if there is a need, it makes sense to collect finer granularity latency with the record timestamp provided in the `Produce` promise. 

Minor suggestion: it would be great to have all related information well commented on in a single place somewhere.

**Comment by twmb** [comment]
Where do you think it could be documented? doc/hooks.md?

**Comment by beanliu** [comment]
> Where do you think it could be documented? doc/hooks.md? doc/monitoring.md?

They both look good to me as long as it is reachable from a related section of the https://github.com/twmb/franz-go page.

**Comment by twmb** [comment]
Sorry about the delay -- v1.4 _should be_ this week. I think there's _one_ more thing to add to it. I'll reopen this for now so that it can be properly closed when v1.4 is released (now that I have a v1.4 tracking issue).

**Comment by twmb** [comment]
https://github.com/twmb/franz-go/commit/30b9f1f3b314a305c6a51788eb75f7d1212c1eb4

To be merged into master soon.

**Comment by twmb** [comment]
Merged!
